src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala

```diff
@@ -0,0 +1,220 @@
+    val nameToAddFile = generateCandidateFileMap(deltaLog.dataPath, candidateFiles)
+
+    scanTimeMs = (System.nanoTime() - startTime) / 1000 / 1000
+
+    val actions: Seq[Action] = if (candidateFiles.isEmpty) {
+      // Case 1: Do nothing if no row qualifies the partition predicates
+      // that are part of Update condition
+      Nil
+    } else if (dataPredicates.isEmpty) {
+      // Case 2: Update all the rows from the files that are in the specified partitions
+      // when the data filter is empty
+      numTouchedFiles = candidateFiles.length
+
+      val filesToRewrite = candidateFiles.map(_.path)
+      val operationTimestamp = System.currentTimeMillis()
+      val deleteActions = candidateFiles.map(_.removeWithTimestamp(operationTimestamp))
+
+      val rewrittenFiles = rewriteFiles(sparkSession, txn, tahoeFileIndex.path,
+        filesToRewrite, nameToAddFile, updateCondition)
+
+      numRewrittenFiles = rewrittenFiles.size
+      rewriteTimeMs = (System.nanoTime() - startTime) / 1000 / 1000 - scanTimeMs
+
+      deleteActions ++ rewrittenFiles
+    } else {
+      // Case 3: Find all the affected files using the user-specified condition
+      val fileIndex = new TahoeBatchFileIndex(
+        sparkSession, "update", candidateFiles, deltaLog, tahoeFileIndex.path, txn.snapshot)
+      // Keep everything from the resolved target except a new TahoeFileIndex
+      // that only involves the affected files instead of all files.
+      val newTarget = DeltaTableUtils.replaceFileIndex(target, fileIndex)
+      val data = Dataset.ofRows(sparkSession, newTarget)
+      val filesToRewrite =
+        withStatusCode("DELTA", s"Finding files to rewrite for UPDATE operation") {
+          data.filter(new Column(updateCondition)).select(input_file_name())
+            .distinct().as[String].collect()
+        }
+
+      scanTimeMs = (System.nanoTime() - startTime) / 1000 / 1000
+      numTouchedFiles = filesToRewrite.length
+
+      if (filesToRewrite.isEmpty) {
+        // Case 3.1: Do nothing if no row qualifies the UPDATE condition
+        Nil
+      } else {
+        // Case 3.2: Delete the old files and generate the new files containing the updated
+        // values
+        val operationTimestamp = System.currentTimeMillis()
+        val deleteActions =
+          removeFilesFromPaths(deltaLog, nameToAddFile, filesToRewrite, operationTimestamp)
+        val rewrittenFiles =
+          withStatusCode("DELTA", s"Rewriting ${filesToRewrite.size} files for UPDATE operation") {
+            rewriteFiles(sparkSession, txn, tahoeFileIndex.path,
+              filesToRewrite, nameToAddFile, updateCondition)
+          }
+
+        numRewrittenFiles = rewrittenFiles.size
+        rewriteTimeMs = (System.nanoTime() - startTime) / 1000 / 1000 - scanTimeMs
+
+        deleteActions ++ rewrittenFiles
+      }
+    }
+
+    if (actions.nonEmpty) {
+      txn.commit(actions, DeltaOperations.Update(condition.map(_.simpleString(5).toString)))
```

Comment:
I think we can just use `toString` here.

Reply:
done.